### PR TITLE
add support for downloading avatars from gravatar

### LINKED ISSUE #1905: support for gravatar profile photo lookups
the social provider/service feature is a great addition to this app.

supporting gravatars (with the same UX as other social providers like twitter or mastadon) would help fill in the gaps for contacts that only have an email address

### PR DESCRIPTION
**Pull request by eleith**
adding support for gravatar to the current social providers/services.
because gravatar relies on email instead of the x-socialprofile carddav
field, this requires re-working the current abstraction for how social
providers work.

also in this change is improved support for when a contact card has
multiple fields of the same type (as might be common with email) thus
they all can be checked for profile photos if one of them doesn't result
in a valid photo

fixes #1905

### CODE REVIEW

**Comment by eleith** [comment]
this is a draft (no linting, no tests), based on a WIP branch i'm using locally to improve photo coverage in my address book.

because gravatar requires the email field, i found the current abstraction layer for the social providers needed some iteration to be more flexible about what parts of a contact card are needed to do photo lookups.

also, because gravatar relies on the email field, i found it quite common that a contact card has multiple email fields, but only one of those would have a gravatar profile photo associated with it. thus, supporting multiple lookups (as opposed to just using the first one, irrespective of if it returns a photo or not) was important.

i've modified the social provider abstraction to both support being more general about what fields are needed for the profile photo lookup and to also support looking up an array of the same fields, if the first one doesn't return a valid profile photo.

is their appetite to see this feature land? if so, i'ld be happy to invest time in improving code quality if i can get some guidance on some of the decisions made currently and high level ideas about what to test.

**Comment by SuperSandro2000** [comment]
Can you also add support for libravatar? I think it uses the same URL scheme but a different domain and is free-er.

File: lib/Service/Social/GravatarProvider.php
Comment (call-me-matt) [inline lib/Service/Social/GravatarProvider.php:5]: credit goes to you!

File: lib/Service/Social/MastodonProvider.php
Comment (call-me-matt) [inline lib/Service/Social/MastodonProvider.php:34]: typo, mastodon

**Review by call-me-matt — COMMENTED** [review]
This looks great! I am very happy to see this feature evolving.
Thank you also for finding and fixing some errors in this PR.
I only found a typo and did not yet have the chance to try it out, but the code looks good.

**Review by skjnldsv — COMMENTED** [review]
Please check the indentation too :)

**Comment by eleith** [comment]
adopted linting/style, addressed some minor feedback and will begin working on unit testing

**Comment by call-me-matt** [comment]
Just in case someone else also does not have any contacts using gravatar, you can use this email address for testing: wzueyuqgwqsgeiiqsr@niwghx.online

**Comment by eleith** [comment]
@call-me-matt code coverage dropped because off pushing more of the `x-socialprofile` logic directly into the individual social providers, duplicating that logic into each of the providers, and also providing handling for supporting multiple image URLs from one provider being found.

there aren't any existing unit tests for the social providers, so i haven't taken a stab at those yet. with that said, if there was an obvious place to abstract some of the `x-socialprofile` handling, it would help decrease the complexity and make testing easier in the future.

otherwise, linting and unit tests are now done (and test coverage of `socialapiservice` actually increased).

File: lib/Service/Social/DiasporaProvider.php
Comment (call-me-matt) [inline lib/Service/Social/DiasporaProvider.php:153]: I guess this needs an \ as well.

File: lib/Service/Social/DiasporaProvider.php
Comment (call-me-matt) [inline lib/Service/Social/DiasporaProvider.php:37]: did not debug, but the diaspora provider seems to be broken. tried with `podmin@joindiaspora.com `, which works on master.

**Review by call-me-matt — CHANGES_REQUESTED** [review]
Great job, thank you

File: lib/Service/Social/DiasporaProvider.php
Comment (eleith) [reply lib/Service/Social/DiasporaProvider.php:37]: after adding unit tests and getting them to work, i can now confirm that the username you provided results in a successful profile image download

**Comment by eleith** [comment]
@call-me-matt all feedback has been addressed. given there was a bug in one of the providers, i went ahead and added unit tests for all of the other providers and this PR should increase code coverage overall now.

**Review by call-me-matt — APPROVED** [review]
looks good, works nicely, great job!

**Comment by eleith** [comment]
@call-me-matt can i merge this myself, or does someone else need to? (the merge button does highlight for me, but wanted to double check...)

**Comment by call-me-matt** [comment]
> can i merge this myself, or does someone else need to?

I guess so, but usually I wait for @skjnldsv as the maintainer of this app for feedback. Normally, he also does the merge right away.

**Comment by skjnldsv** [comment]
> > can i merge this myself, or does someone else need to?
> 
> I guess so, but usually I wait for @skjnldsv as the maintainer of this app for feedback. Normally, he also does the merge right away.

If it's green, please do! 💪✨

**Comment by eleith** [comment]
thanks everyone. pleasure working with you both on this one.

**Pull request merged**
